### PR TITLE
Move to GH Actions

### PR DESCRIPTION
**Pull request by sabderemane**
Related to #344

### CODE REVIEW

**Review by ezio-melotti — COMMENTED** [review]
An early review with a few comments.

**Comment by arhadthedev** [comment]
@sabderemane Would you mind to address the reviews, please? It would be great to retire a third party Heroku service of extra maintainance, not-so-transparent deployment, non-public error logs, and a dedicated, actively waiting server loop.

**Comment by sabderemane** [comment]
I will, for sure, I just didn't have time to do so yet :)

File: .github/workflows/review_pr.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('pyproject.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
```
[the remark this answers is not in the document]
Comment (sabderemane) [reply .github/workflows/review_pr.yml:23]: I update the `key` but for the rest it should be ok, I use [this example](https://github.com/actions/cache/blob/main/examples.md#simple-example)

File: .github/workflows/review_pr.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      - uses: actions/setup-python@v3
+        with:
+          python-version: '3.9'
```
[the remark this answers is not in the document]
Comment (sabderemane) [reply .github/workflows/review_pr.yml]: No, I changed

File: bedevere/backport.py [hunk 3]
```diff
@@ -41,68 +49,63 @@ async def _remove_backport_label(gh, original_issue, branch, backport_pr_number)
...
-        return
-    pull_request = event.data["pull_request"]
-    title = util.normalize_title(pull_request['title'],
+    response = await gh.post(original_issue['comments_url'], data={'body': message})
+    return response
+
+async def manage_labels(gh, *args, **kwargs):
+    with open(os.environ["GITHUB_EVENT_PATH"]) as f:
+        event = json.load(f)
+        if event.get("action") == "edited" and "title" not in event.get("changes"):
```
[the remark this answers is not in the document]
Comment (sabderemane) [reply bedevere/backport.py]: No, I only keep `event` only, changed

File: bedevere/review_pr.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    try:
+        async with aiohttp.ClientSession() as session:
+            gh = GitHubAPI(session, "sabderemane", oauth_token=os.getenv("GH_AUTH"))
+            await dismiss_invalid_pr_review_request(gh)  
+    except Exception:
+        traceback.print_exc()
+
+
+loop = asyncio.get_event_loop()
+loop.run_until_complete(main())
```
[the remark this answers is not in the document]
Comment (sabderemane) [reply bedevere/review_pr.py]: updated

File: .github/workflows/review_pr.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+on:
+  pull_request:
+    types: [review_requested]
+
+env:
+  GH_AUTH: ${{ secrets.GH_AUTH }}
+
+jobs:
+  review_invalid_pr:
+    name: Check and update backport label of PR
```
[the remark this answers is not in the document]
Comment (sabderemane) [reply .github/workflows/review_pr.yml]: Updated

**Review by sabderemane — COMMENTED** [review]
Thank you @ezio-melotti for the review 👍🏽

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -1,83 +1,18 @@
 Contributing and Maintenance Guide
 ==================================
 
-Bedevere web service is deployed to Heroku, which is managed by The PSF.
+Bedevere web service is currently launched with GitHub Actions, which is managed by The PSF.
```
Comment (hugovk) [inline CONTRIBUTING.rst:4]: "currently" is redundant
```suggestion
Bedevere web service is launched with GitHub Actions, which is managed by The PSF.
```

File: .github/workflows/review_pr.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (hugovk) [reply .github/workflows/review_pr.yml:23]: `actions/setup-python` now natively includes pip caching.

Docs: https://github.com/actions/setup-python#caching-packages-dependencies

So we can remove the `uses: actions/cache@v1` step, and update:


```diff
-      - uses: actions/setup-python@v3
+      - uses: actions/setup-python@v4
         with:
           python-version: '3.x'
+          cache: pip
+          cache-dependency-path: '*requirements.txt'
```

And the same in the other workflows.

File: .github/workflows/review_pr.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  GH_AUTH: ${{ secrets.GH_AUTH }}
+
+jobs:
+  review_invalid_pr:
+    name: Dismiss review request from the invalid PR
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (hugovk) [inline .github/workflows/review_pr.yml:17]: Bump:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/close_pr.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  GH_AUTH: ${{ secrets.GH_AUTH }}
+
+jobs:
+  close_pr:
+    name: close invalid PR
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (hugovk) [inline .github/workflows/close_pr.yml:17]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/close_pr.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - uses: actions/setup-python@v3
```
Comment (hugovk) [inline .github/workflows/close_pr.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

File: .github/workflows/backport.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  GH_AUTH: ${{ secrets.GH_AUTH }}
+
+jobs:
+  backport:
+    name: Check and update backport label of PR
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (hugovk) [inline .github/workflows/backport.yml:17]: Bump version:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/backport.yml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - uses: actions/setup-python@v3
```
Comment (hugovk) [inline .github/workflows/backport.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

File: .github/workflows/filepaths.yml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  GH_AUTH: ${{ secrets.GH_AUTH }}
+
+jobs:
+  filepaths:
+    name: Checks filepaths on a PR
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (hugovk) [inline .github/workflows/filepaths.yml:17]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/filepaths.yml [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - uses: actions/setup-python@v3
```
Comment (hugovk) [inline .github/workflows/filepaths.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

**Review by hugovk — COMMENTED** [review]
A few workflow suggestions :)

File: .github/workflows/close_pr.yml [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+      - uses: actions/checkout@v2
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('pyproject.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - uses: actions/setup-python@v3
+        with:
+          python-version: '3.9'
```
[the remark this answers is not in the document]
Comment (hugovk) [reply .github/workflows/close_pr.yml]: Dependency path: we install `dev-requirements.txt`, which includes installing `-r requirements.txt`

below I suggest using `cache-dependency-path: '*requirements.txt'`, so if either of those files change, it invalidates the cache and downloads the new things.

File: bedevere/backport.py [hunk 15]
```diff
@@ -1,34 +1,42 @@
...
 async def _copy_over_labels(gh, original_issue, backport_issue):
     """Copy over relevant labels from the original PR to the backport PR."""
     label_prefixes = "skip", "type", "sprint"
     labels = list(filter(lambda x: x.startswith(label_prefixes),
                     util.labels(original_issue)))
     if labels:
-        await gh.post(backport_issue["labels_url"], data=labels)
+        response = await gh.post(backport_issue["labels_url"], data=labels)
+        return response
+    return "no labels"
```
Comment (ezio-melotti) [inline bedevere/backport.py:38]: Why were the return values added?  ISTM that they are not used.

File: bedevere/backport.py [hunk 16]
```diff
@@ -41,83 +49,63 @@ async def _remove_backport_label(gh, original_issue, branch, backport_pr_number)
...
-
-    >>> is_maintenance_branch("3.11")
-    True
-    >>> is_maintenance_branch("main")
-    False
-    >>> is_maintenance_branch("gh-1234/something-completely-different")
-    False
-    """
-    maintenance_branch_pattern = r'\d+\.\d+'
-    return bool(re.fullmatch(maintenance_branch_pattern, ref))
```
Comment (ezio-melotti) [inline bedevere/backport.py:82]: I think this got accidentally deleted during the merge (like the `re.IGNORECASE` above).

File: bedevere/backport.py [hunk 17]
```diff
@@ -41,83 +49,63 @@ async def _remove_backport_label(gh, original_issue, branch, backport_pr_number)
...
+    with open(os.environ["GITHUB_EVENT_PATH"]) as f:
+        event = json.load(f)
+    if event.get("action") == "edited" and "title" not in event.get("changes"):
         return
-    pull_request = event.data["pull_request"]
+    pull_request = event["pull_request"]
     base_branch = pull_request["base"]["ref"]
 
-    if not is_maintenance_branch(base_branch):
+    if base_branch == "main":
```
Comment (ezio-melotti) [inline bedevere/backport.py:92]: This should also use `not is_maintanance_branch`.
You should double check that there aren't other updates that got revert during the merge (or possibly re-do the merge).

File: bedevere/close_pr.py [hunk 18]
```diff
@@ -1,61 +1,56 @@
...
     PR is considered invalid if:
     * base_label is 'python:main'
     * head_label is '<username>:<maint_branch>'
     """
-    head_label = event.data["pull_request"]["head"]["label"]
-    base_label = event.data["pull_request"]["base"]["label"]
+    with open(os.environ["GITHUB_EVENT_PATH"]) as f:
+        event = json.load(f)
+        head_label = event["pull_request"]["head"]["label"]
+        base_label = event["pull_request"]["base"]["label"]
```
Comment (ezio-melotti) [inline bedevere/close_pr.py:31]:
```suggestion
    head_label = event["pull_request"]["head"]["label"]
    base_label = event["pull_request"]["base"]["label"]
```

File: bedevere/filepaths.py [hunk 19]
```diff
@@ -1,22 +1,35 @@
...
-
-@router.register('pull_request', action='opened')
-@router.register('pull_request', action='synchronize')
-@router.register('pull_request', action='reopened')
-async def check_file_paths(event, gh, *args, **kwargs):
-    pull_request = event.data['pull_request']
+async def check_file_paths(gh, *args, **kwargs):
+    with open(os.environ["GITHUB_EVENT_PATH"]) as f:
+        event = json.load(f)
+        pull_request = event['pull_request']
```
Comment (ezio-melotti) [inline bedevere/filepaths.py:18]:
```suggestion
    pull_request = event['pull_request']
```

File: bedevere/review_pr.py [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    """Dismiss review request from the invalid PR.
+
+    PR is considered invalid if:
+    * base_label is 'python:main'
+    * head_label is '<username>:<maint_branch>'
+    """
+    with open(os.environ["GITHUB_EVENT_PATH"]) as f:
+        event = json.load(f)
+        head_label = event["pull_request"]["head"]["label"]
+        base_label = event["pull_request"]["base"]["label"]
```
Comment (ezio-melotti) [inline bedevere/review_pr.py:34]:
```suggestion
    head_label = event["pull_request"]["head"]["label"]
    base_label = event["pull_request"]["base"]["label"]
```

File: bedevere/review_pr.py [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+    if PYTHON_MAINT_BRANCH_RE.match(head_label) and \
+            base_label == "python:main":
+        data = {"reviewers": [reviewer["login"] for reviewer in event["pull_request"]["requested_reviewers"]],
+                "team_reviewers": [team["name"] for team in event["pull_request"]["requested_teams"]]
+                }
+        await gh.delete(f'{event["pull_request"]["url"]}/requested_reviewers',
+                        data=data)
```
Comment (ezio-melotti) [inline bedevere/review_pr.py:42]:
```suggestion
    if (PYTHON_MAINT_BRANCH_RE.match(head_label) and
        base_label == "python:main"):
        pr = event["pull_request"]
        reviewers = [reviewer["login"] for reviewer in pr["requested_reviewers"]
        team_reviewers = [team["name"] for team in pr["requested_teams"]
        await gh.delete(f'{event["pull_request"]["url"]}/requested_reviewers',
                        data=dict(reviewers=reviewers, team_reviewers=team_reviewers))
```

**Review by ezio-melotti — CHANGES_REQUESTED** [review]
I left some more review comments.
* It seems that some of the changes applied lately got lost during a merge -- those should be double checked and fixed;
* there were some formatting problems likely caused by a misconfigured editor (trailing spaces, tabs, changed indentation)

P.S.: you can probably batch merge most of these directly from GitHub.

**Comment by sabderemane** [comment]
Sorry for the late update, I address all the issues hopefully, thank you for the review @hugovk, @ezio-melotti and @DanielNoord.  I change my configuration to avoid editor issue 🙏🏽  I will see to continue and update tests.

**Comment by arhadthedev** [comment]
From https://github.com/python/bedevere/issues/344#issuecomment-1591861785:

> One issue with using Actions is that it require maintainers to "approve" the workflow for contributors before it can be run.
The workflows done by bedevere, like checking for skip issue, skip news, applying "type" labels, should just run by itself without needing maintainer approval.
>
> So I think it would be better to convert bedevere into GitHub App instead of Actions. By converting this to GitHub App, we can resolve the issue where 2FA is needed for the bedevere-bot account.

From the following https://github.com/python/bedevere/issues/344#issuecomment-1591889387:

> To solve the issue where core devs want to be able to re-run workflows/retrigger webhook in case the web app is down, we can grant the "[GitHub App manager](https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/adding-and-removing-github-app-managers-in-your-organization#about-github-app-managers)" access to core devs.

The problem surfaced while working on gh-568 so gh-569 was created instead.

@sabderemane Probably, your PR is a dead-end then. We need to either close it in favor of gh-569 or think on some alterations.

**Comment by sabderemane** [comment]
Closing in favor of gh-569, happy to help on other things instead :)

**Pull request closed without merging**